src/backend/distributed/planner/multi_logical_planner.c

```diff
@@ -186,6 +207,93 @@ MultiLogicalPlanCreate(Query *originalQuery, Query *queryTree,
 
 
 /*
+ * ReplaceExternParamsInOriginalQuery replaces the external parameters that appears
+ * in the query with the corresponding entries in the boundParams.
+ *
+ * Note that this function is inspired by eval_const_expr() on Postgres. We cannot
+ * use that function because it requires access to PlannerInfo.
+ */
+static Node *
+ReplaceExternParamsInOriginalQuery(Node *inputNode, ParamListInfo boundParams)
+{
+	int parameterIndex = 0;
+
+	Assert(boundParams != NULL);
+
+	if (inputNode == NULL)
+	{
+		return NULL;
+	}
+
+	if (IsA(inputNode, Param))
+	{
+		Param *paramToProcess = (Param *) inputNode;
+		ParamExternData correspondingParameterData;
+		int numberOfParameters = boundParams->numParams;
+		int parameterId = paramToProcess->paramid;
+		int16 typeLength = 0;
+		bool typeByValue = false;
+		Datum constValue = 0;
+		bool paramIsNull = false;
+
+		if (paramToProcess->paramkind != PARAM_EXTERN)
+		{
+			return inputNode;
+		}
+
+		if (parameterId < 0)
+		{
+			return inputNode;
+		}
+
+		/* parameterId starts from 1 */
+		parameterIndex = parameterId - 1;
+		if (parameterIndex >= numberOfParameters)
+		{
+			return inputNode;
+		}
+
+		correspondingParameterData = boundParams->params[parameterIndex];
+
+		if (!(correspondingParameterData.pflags & PARAM_FLAG_CONST))
+		{
+			return inputNode;
+		}
+
+		get_typlenbyval(paramToProcess->paramtype, &typeLength, &typeByValue);
+
+		paramIsNull = correspondingParameterData.isnull;
+		if (paramIsNull)
+		{
+			constValue = 0;
+		}
+		else if (typeByValue)
+		{
+			constValue = correspondingParameterData.value;
+		}
+		else
+		{
+			constValue = datumCopy(correspondingParameterData.value, typeByValue,
```

Comment:
Hm. Actually. Why do we need to copy the datum?  We can just reference it in the const I think?